Mxnet test case failed

**Describe the bug**

tests/test_mxnet.py  

add test `test_byteps_push_pull_inplace`

```python
if __name__ == '__main__':
    mxtest = MXTest()
    bps.init()
    mxtest.test_byteps_push_pull()
    mxtest.test_byteps_trainer_param_order()
    #mxtest.test_byteps_broadcast()
    mxtest.test_byteps_push_pull_inplace() #SEE HERE 
    # TODO: fix shutdown issue https://github.com/bytedance/byteps/issues/177
    # bps.shutdown()
```

```sh
export BYTEPS_LOG_LEVEL=DEBUG
export BYTEPS_FORCE_DISTRIBUTED=1
export DMLC_ROLE=worker
export DMLC_PS_ROOT_URI=xxx.xxx.xxx.xxx
export DMLC_PS_ROOT_PORT=1234
export DMLC_WORKER_ID=0
export DMLC_NUM_WORKER=1
export NVIDIA_VISIBLE_DEVICES=0,1,2,3
```

And I will get this error: 
```
Traceback (most recent call last):
    File "/usr/local/byteps/tests/test_mxnet.py", line 189, in <module>
        mxtest.test_byteps_push_pull_inplace()
    File "/usr/local/byteps/tests/test_mxnet.py", line 134, in test_byteps_push_pull_inplace
AssertionError: bps.byteps_push_pull procuces. 
     incorrect results for self
```

log
```
BytePS launching worker

!!!Enable profiling for WORKER_ID: 0 and local_rank: 0!!!
BYTEPS_TRACE_START_STEP: 	BYTEPS_TRACE_END_STEP: 	 BYTEPS_TRACE_DIR: 
Command: python3 /usr/local/byteps/tests/test_mxnet.py


!!!Enable profiling for WORKER_ID: 0 and local_rank: 1!!!
BYTEPS_TRACE_START_STEP: 	BYTEPS_TRACE_END_STEP: 	 BYTEPS_TRACE_DIR: 
Command: python3 /usr/local/byteps/tests/test_mxnet.py


!!!Enable profiling for WORKER_ID: 0 and local_rank: 2!!!
BYTEPS_TRACE_START_STEP: 	BYTEPS_TRACE_END_STEP: 	 BYTEPS_TRACE_DIR: 
Command: python3 /usr/local/byteps/tests/test_mxnet.py


!!!Enable profiling for WORKER_ID: 0 and local_rank: 3!!!
BYTEPS_TRACE_START_STEP: 	BYTEPS_TRACE_END_STEP: 	 BYTEPS_TRACE_DIR: 
Command: python3 /usr/local/byteps/tests/test_mxnet.py

[2020-01-13 06:36:23.776870: D byteps/common/communicator.cc:63] Using Communicator=Socket
[2020-01-13 06:36:23.776992: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_3
[2020-01-13 06:36:23.777035: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_3
[2020-01-13 06:36:23.777080: D byteps/common/communicator.cc:121] This is ROOT device, rank=3, all sockets create successfully
[2020-01-13 06:36:23.777094: D byteps/common/global.cc:124] Partition size round up to 4096000 (bytes)
[2020-01-13 06:36:23.777103: D byteps/common/global.cc:140] Using key hash function type: djb2
[2020-01-13 06:36:23.777112: D byteps/common/global.cc:151] Number of worker=1, launching distributed job
[2020-01-13 06:36:23.777128: D byteps/common/nccl_manager.cc:133] nccl_group_size set to 4
[2020-01-13 06:36:23.777137: D byteps/common/nccl_manager.cc:152] nccl_pcie_size set to 4
[2020-01-13 06:36:23.777144: D byteps/common/nccl_manager.cc:154] nccl_pcie_num set to 1
[2020-01-13 06:36:23.777147: D byteps/common/communicator.cc:164] Listening on socket 3
[2020-01-13 06:36:23.777219: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_nccl3
[2020-01-13 06:36:23.777257: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_nccl3
[2020-01-13 06:36:23.777296: D byteps/common/communicator.cc:55] This is nccl ROOT device, rank=3, all sockets create successfully
[2020-01-13 06:36:23.777308: D byteps/common/nccl_manager.cc:85] Constructing NCCL communicators. 0 1 2 3
[2020-01-13 06:36:23.777325: D byteps/common/communicator.cc:164] Listening on socket 3
[2020-01-13 06:36:24.284394: D byteps/common/communicator.cc:63] Using Communicator=Socket
[2020-01-13 06:36:24.284588: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_1
[2020-01-13 06:36:24.284657: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_1
[2020-01-13 06:36:24.284678: D byteps/common/communicator.cc:121] This is WORKER device, rank=1, all sockets create successfully
[2020-01-13 06:36:24.284696: D byteps/common/global.cc:124] Partition size round up to 4096000 (bytes)
[2020-01-13 06:36:24.284715: D byteps/common/global.cc:140] Using key hash function type: djb2
[2020-01-13 06:36:24.284731: D byteps/common/global.cc:151] Number of worker=1, launching distributed job
[2020-01-13 06:36:24.284762: D byteps/common/nccl_manager.cc:133] nccl_group_size set to 4
[2020-01-13 06:36:24.284779: D byteps/common/nccl_manager.cc:152] nccl_pcie_size set to 4
[2020-01-13 06:36:24.284795: D byteps/common/nccl_manager.cc:154] nccl_pcie_num set to 1
[2020-01-13 06:36:24.284864: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_nccl1
[2020-01-13 06:36:24.284923: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_nccl1
[2020-01-13 06:36:24.284943: D byteps/common/communicator.cc:55] This is nccl WORKER device, rank=1, all sockets create successfully
[2020-01-13 06:36:24.284958: D byteps/common/nccl_manager.cc:85] Constructing NCCL communicators. 0 1 2 3
[2020-01-13 06:36:24.367309: D byteps/common/communicator.cc:63] Using Communicator=Socket
[2020-01-13 06:36:24.367538: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_0
[2020-01-13 06:36:24.367611: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_0
[2020-01-13 06:36:24.367649: D byteps/common/communicator.cc:121] This is WORKER device, rank=0, all sockets create successfully
[2020-01-13 06:36:24.367672: D byteps/common/global.cc:124] Partition size round up to 4096000 (bytes)
[2020-01-13 06:36:24.367692: D byteps/common/global.cc:140] Using key hash function type: djb2
[2020-01-13 06:36:24.367711: D byteps/common/global.cc:151] Number of worker=1, launching distributed job
[2020-01-13 06:36:24.367743: D byteps/common/nccl_manager.cc:133] nccl_group_size set to 4
[2020-01-13 06:36:24.367763: D byteps/common/nccl_manager.cc:152] nccl_pcie_size set to 4
[2020-01-13 06:36:24.367780: D byteps/common/nccl_manager.cc:154] nccl_pcie_num set to 1
[2020-01-13 06:36:24.367867: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_nccl0
[2020-01-13 06:36:24.367930: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_nccl0
[2020-01-13 06:36:24.367950: D byteps/common/communicator.cc:55] This is nccl WORKER device, rank=0, all sockets create successfully
[2020-01-13 06:36:24.367967: D byteps/common/nccl_manager.cc:85] Constructing NCCL communicators. 0 1 2 3
[2020-01-13 06:36:24.459078: D byteps/common/communicator.cc:63] Using Communicator=Socket
[2020-01-13 06:36:24.459305: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_2
[2020-01-13 06:36:24.459383: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_2
[2020-01-13 06:36:24.459405: D byteps/common/communicator.cc:121] This is WORKER device, rank=2, all sockets create successfully
[2020-01-13 06:36:24.459424: D byteps/common/global.cc:124] Partition size round up to 4096000 (bytes)
[2020-01-13 06:36:24.459444: D byteps/common/global.cc:140] Using key hash function type: djb2
[2020-01-13 06:36:24.459462: D byteps/common/global.cc:151] Number of worker=1, launching distributed job
[2020-01-13 06:36:24.459497: D byteps/common/nccl_manager.cc:133] nccl_group_size set to 4
[2020-01-13 06:36:24.459516: D byteps/common/nccl_manager.cc:152] nccl_pcie_size set to 4
[2020-01-13 06:36:24.459532: D byteps/common/nccl_manager.cc:154] nccl_pcie_num set to 1
[2020-01-13 06:36:24.459607: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_send_nccl2
[2020-01-13 06:36:24.459674: D byteps/common/communicator.cc:157] Init socket at /tmp/socket_recv_nccl2
[2020-01-13 06:36:24.459694: D byteps/common/communicator.cc:55] This is nccl WORKER device, rank=2, all sockets create successfully
[2020-01-13 06:36:24.459711: D byteps/common/nccl_manager.cc:85] Constructing NCCL communicators. 0 1 2 3
[2020-01-13 06:36:27.476497: D byteps/common/nccl_manager.cc:104] root nccl_id is 1458356680124661762
[2020-01-13 06:36:27.864698: D byteps/common/nccl_manager.cc:114] recv nccl_id is 1458356680124661762, local_rank=2
[2020-01-13 06:36:27.965359: D byteps/common/nccl_manager.cc:114] recv nccl_id is 1458356680124661762, local_rank=1
[2020-01-13 06:36:28.124460: D byteps/common/nccl_manager.cc:114] recv nccl_id is 1458356680124661762, local_rank=0
[2020-01-13 06:36:28.922971: D byteps/common/global.cc:226] Create schedule queue 0
[2020-01-13 06:36:28.923055: D byteps/common/global.cc:226] Create schedule queue 1
[2020-01-13 06:36:28.923075: D byteps/common/global.cc:226] Create schedule queue 2
[2020-01-13 06:36:28.923090: D byteps/common/global.cc:226] Create schedule queue 3
[2020-01-13 06:36:28.923105: D byteps/common/global.cc:226] Create schedule queue 4
[2020-01-13 06:36:28.923119: D byteps/common/global.cc:226] Create schedule queue 5
[2020-01-13 06:36:28.923134: D byteps/common/global.cc:226] Create schedule queue 6
[2020-01-13 06:36:28.923149: D byteps/common/global.cc:226] Create schedule queue 7
[2020-01-13 06:36:28.923164: D byteps/common/global.cc:226] Create schedule queue 8
[2020-01-13 06:36:28.923178: D byteps/common/global.cc:226] Create schedule queue 9
[2020-01-13 06:36:28.923195: D byteps/common/global.cc:234] Inited rank=2 local_rank=2 size=4 local_size=4 worker_id=0
[2020-01-13 06:36:28.923744: D byteps/common/global.cc:266] Started 8 background threads. rank=2
[2020-01-13 06:36:28.927974: D byteps/common/global.cc:226] Create schedule queue 0
[2020-01-13 06:36:28.928035: D byteps/common/global.cc:226] Create schedule queue 1
[2020-01-13 06:36:28.928054: D byteps/common/global.cc:226] Create schedule queue 2
[2020-01-13 06:36:28.928070: D byteps/common/global.cc:226] Create schedule queue 3
[2020-01-13 06:36:28.928095: D byteps/common/global.cc:226] Create schedule queue 4
[2020-01-13 06:36:28.928121: D byteps/common/global.cc:226] Create schedule queue 5
[2020-01-13 06:36:28.928141: D byteps/common/global.cc:226] Create schedule queue 6
[2020-01-13 06:36:28.928118: D byteps/common/global.cc:226] Create schedule queue 0
[2020-01-13 06:36:28.928156: D byteps/common/global.cc:226] Create schedule queue 7
[2020-01-13 06:36:28.928173: D byteps/common/global.cc:226] Create schedule queue 8
[2020-01-13 06:36:28.928175: D byteps/common/global.cc:226] Create schedule queue 1
[2020-01-13 06:36:28.928188: D byteps/common/global.cc:226] Create schedule queue 9
[2020-01-13 06:36:28.928197: D byteps/common/global.cc:226] Create schedule queue 2
[2020-01-13 06:36:28.928206: D byteps/common/global.cc:234] Inited rank=3 local_rank=3 size=4 local_size=4 worker_id=0
[2020-01-13 06:36:28.928214: D byteps/common/global.cc:226] Create schedule queue 3
[2020-01-13 06:36:28.928231: D byteps/common/global.cc:226] Create schedule queue 4
[2020-01-13 06:36:28.928246: D byteps/common/global.cc:226] Create schedule queue 5
[2020-01-13 06:36:28.928261: D byteps/common/global.cc:226] Create schedule queue 6
[2020-01-13 06:36:28.928276: D byteps/common/global.cc:226] Create schedule queue 7
[2020-01-13 06:36:28.928291: D byteps/common/global.cc:226] Create schedule queue 8
[2020-01-13 06:36:28.928307: D byteps/common/global.cc:226] Create schedule queue 9
[2020-01-13 06:36:28.928324: D byteps/common/global.cc:234] Inited rank=1 local_rank=1 size=4 local_size=4 worker_id=0
[2020-01-13 06:36:28.928581: D byteps/common/global.cc:266] Started 6 background threads. rank=3
[2020-01-13 06:36:28.928745: D byteps/common/global.cc:266] Started 8 background threads. rank=1
[2020-01-13 06:36:28.930599: D byteps/common/global.cc:226] Create schedule queue 0
[2020-01-13 06:36:28.930674: D byteps/common/global.cc:226] Create schedule queue 1
[2020-01-13 06:36:28.930710: D byteps/common/global.cc:226] Create schedule queue 2
[2020-01-13 06:36:28.930728: D byteps/common/global.cc:226] Create schedule queue 3
[2020-01-13 06:36:28.930746: D byteps/common/global.cc:226] Create schedule queue 4
[2020-01-13 06:36:28.930763: D byteps/common/global.cc:226] Create schedule queue 5
[2020-01-13 06:36:28.930780: D byteps/common/global.cc:226] Create schedule queue 6
[2020-01-13 06:36:28.930797: D byteps/common/global.cc:226] Create schedule queue 7
[2020-01-13 06:36:28.930828: D byteps/common/global.cc:226] Create schedule queue 8
[2020-01-13 06:36:28.930838: D byteps/common/global.cc:226] Create schedule queue 9
[2020-01-13 06:36:28.930850: D byteps/common/global.cc:234] Inited rank=0 local_rank=0 size=4 local_size=4 worker_id=0
[2020-01-13 06:36:28.931318: D byteps/common/global.cc:266] Started 8 background threads. rank=0
[2020-01-13 06:36:29.856225: D byteps/common/global.cc:353] Declared tensor byteps.tensor_100, declared key (not PS key): 0 rank=2
[2020-01-13 06:36:29.856402: D byteps/common/operations.cc:247] byteps.tensor_100 partitioned to 1 part(s), total_len=68, key_range=[0, 0] rank=2
[2020-01-13 06:36:29.991027: D byteps/common/global.cc:353] Declared tensor byteps.tensor_100, declared key (not PS key): 0 rank=0
[2020-01-13 06:36:29.991128: D byteps/common/operations.cc:247] byteps.tensor_100 partitioned to 1 part(s), total_len=68, key_range=[0, 0] rank=0
[2020-01-13 06:36:30. 44476: D byteps/common/global.cc:353] Declared tensor byteps.tensor_100, declared key (not PS key): 0 rank=1
[2020-01-13 06:36:30. 44627: D byteps/common/operations.cc:247] byteps.tensor_100 partitioned to 1 part(s), total_len=68, key_range=[0, 0] rank=1
[2020-01-13 06:36:30. 72022: D byteps/common/global.cc:353] Declared tensor byteps.tensor_100, declared key (not PS key): 0 rank=3
[2020-01-13 06:36:30. 72129: D byteps/common/operations.cc:247] byteps.tensor_100 partitioned to 1 part(s), total_len=68, key_range=[0, 0] rank=3
[06:36:30] src/./zmq_van.h:287: Start ZMQ recv thread
[06:36:30] src/./zmq_van.h:287: Start ZMQ recv thread
[06:36:30] src/./zmq_van.h:287: Start ZMQ recv thread
[2020-01-13 06:36:30.175219: D byteps/common/global.cc:534] key 0 assigned to server 0, accumulated workload for this server is 68
[2020-01-13 06:36:30.199473: D byteps/common/global.cc:353] Declared tensor byteps.parameter_0, declared key (not PS key): 1 rank=3
[2020-01-13 06:36:30.199546: D byteps/common/global.cc:353] Declared tensor byteps.gradient_0, declared key (not PS key): 2 rank=3
[2020-01-13 06:36:30.199581: D byteps/common/global.cc:353] Declared tensor byteps.parameter_1, declared key (not PS key): 3 rank=3
[2020-01-13 06:36:30.199620: D byteps/common/global.cc:353] Declared tensor byteps.gradient_1, declared key (not PS key): 4 rank=3
[2020-01-13 06:36:30.199840: D byteps/common/operations.cc:247] byteps.parameter_0 partitioned to 1 part(s), total_len=400, key_range=[65536, 65536] rank=3
[2020-01-13 06:36:30.199865: D byteps/common/operations.cc:267] byteps.parameter_0 is already on cpu, len=400
[2020-01-13 06:36:30.200667: D byteps/common/global.cc:534] key 65536 assigned to server 0, accumulated workload for this server is 468
[2020-01-13 06:36:30.208144: D byteps/common/global.cc:353] Declared tensor byteps.parameter_0, declared key (not PS key): 1 rank=0
[2020-01-13 06:36:30.208218: D byteps/common/global.cc:353] Declared tensor byteps.gradient_0, declared key (not PS key): 2 rank=0
[2020-01-13 06:36:30.208251: D byteps/common/global.cc:353] Declared tensor byteps.parameter_1, declared key (not PS key): 3 rank=0
[2020-01-13 06:36:30.208287: D byteps/common/global.cc:353] Declared tensor byteps.gradient_1, declared key (not PS key): 4 rank=0
[2020-01-13 06:36:30.208331: D byteps/common/operations.cc:247] byteps.parameter_0 partitioned to 1 part(s), total_len=400, key_range=[65536, 65536] rank=0
[2020-01-13 06:36:30.208346: D byteps/common/operations.cc:267] byteps.parameter_0 is already on cpu, len=400
[2020-01-13 06:36:30.219204: D byteps/common/global.cc:353] Declared tensor byteps.parameter_0, declared key (not PS key): 1 rank=1
[2020-01-13 06:36:30.219338: D byteps/common/global.cc:353] Declared tensor byteps.gradient_0, declared key (not PS key): 2 rank=1
[2020-01-13 06:36:30.219444: D byteps/common/global.cc:353] Declared tensor byteps.parameter_1, declared key (not PS key): 3 rank=1
[2020-01-13 06:36:30.219564: D byteps/common/global.cc:353] Declared tensor byteps.gradient_1, declared key (not PS key): 4 rank=1
[2020-01-13 06:36:30.221256: D byteps/common/operations.cc:247] byteps.parameter_0 partitioned to 1 part(s), total_len=400, key_range=[65536, 65536] rank=1
[2020-01-13 06:36:30.221307: D byteps/common/operations.cc:267] byteps.parameter_0 is already on cpu, len=400
[2020-01-13 06:36:30.222320: D byteps/common/global.cc:353] Declared tensor byteps.parameter_0, declared key (not PS key): 1 rank=2
[2020-01-13 06:36:30.222426: D byteps/common/global.cc:353] Declared tensor byteps.gradient_0, declared key (not PS key): 2 rank=2
[2020-01-13 06:36:30.222478: D byteps/common/global.cc:353] Declared tensor byteps.parameter_1, declared key (not PS key): 3 rank=2
[2020-01-13 06:36:30.222535: D byteps/common/global.cc:353] Declared tensor byteps.gradient_1, declared key (not PS key): 4 rank=2
[2020-01-13 06:36:30.224739: D byteps/common/operations.cc:247] byteps.parameter_0 partitioned to 1 part(s), total_len=400, key_range=[65536, 65536] rank=2
[2020-01-13 06:36:30.224797: D byteps/common/operations.cc:267] byteps.parameter_0 is already on cpu, len=400
[2020-01-13 06:36:30.232399: D byteps/common/operations.cc:247] byteps.parameter_1 partitioned to 1 part(s), total_len=400, key_range=[196608, 196608] rank=0
[2020-01-13 06:36:30.232431: D byteps/common/operations.cc:267] byteps.parameter_1 is already on cpu, len=400
[2020-01-13 06:36:30.232446: D byteps/common/operations.cc:247] byteps.parameter_1 partitioned to 1 part(s), total_len=400, key_range=[196608, 196608] rank=1
[2020-01-13 06:36:30.232547: D byteps/common/operations.cc:267] byteps.parameter_1 is already on cpu, len=400
[2020-01-13 06:36:30.232724: D byteps/common/operations.cc:247] byteps.parameter_1 partitioned to 1 part(s), total_len=400, key_range=[196608, 196608] rank=2
[2020-01-13 06:36:30.232775: D byteps/common/operations.cc:267] byteps.parameter_1 is already on cpu, len=400
[2020-01-13 06:36:30.232987: D byteps/common/operations.cc:247] byteps.parameter_1 partitioned to 1 part(s), total_len=400, key_range=[196608, 196608] rank=3
[2020-01-13 06:36:30.233015: D byteps/common/operations.cc:267] byteps.parameter_1 is already on cpu, len=400
[2020-01-13 06:36:30.236672: D byteps/common/global.cc:534] key 196608 assigned to server 0, accumulated workload for this server is 868
[2020-01-13 06:36:30.245615: D byteps/common/global.cc:353] Declared tensor byteps.tensor_200, declared key (not PS key): 5 rank=2
[2020-01-13 06:36:30.245704: D byteps/common/operations.cc:247] byteps.tensor_200 partitioned to 1 part(s), total_len=68, key_range=[327680, 327680] rank=2
[2020-01-13 06:36:30.247157: D byteps/common/global.cc:353] Declared tensor byteps.tensor_200, declared key (not PS key): 5 rank=1
[2020-01-13 06:36:30.247306: D byteps/common/operations.cc:247] byteps.tensor_200 partitioned to 1 part(s), total_len=68, key_range=[327680, 327680] rank=1
[2020-01-13 06:36:30.248285: D byteps/common/global.cc:353] Declared tensor byteps.tensor_200, declared key (not PS key): 5 rank=0
[2020-01-13 06:36:30.248330: D byteps/common/operations.cc:247] byteps.tensor_200 partitioned to 1 part(s), total_len=68, key_range=[327680, 327680] rank=0
[2020-01-13 06:36:30.252102: D byteps/common/global.cc:353] Declared tensor byteps.tensor_200, declared key (not PS key): 5 rank=3
[2020-01-13 06:36:30.252221: D byteps/common/operations.cc:247] byteps.tensor_200 partitioned to 1 part(s), total_len=68, key_range=[327680, 327680] rank=3
[2020-01-13 06:36:30.253121: D byteps/common/global.cc:534] key 327680 assigned to server 0, accumulated workload for this server is 936
tensor before push_pull: 
[ 44.921947  -22.404182   19.133305   13.094497   11.725103  -69.86146
   0.1105547  33.98166    92.52082   -85.41071    16.735744   24.991835
 -77.336075   14.822626   25.756227   43.89869    58.843494 ]
<NDArray 17 @gpu(3)>
tensor after push_pull: 
[ -2.2648907 -26.942665   -3.1381292  30.620129   14.483914  -24.707336
  57.58606    17.05282     9.471867  -14.655745  -26.668665    4.1452713
  29.551579   31.442118   47.59712     5.5366697   6.2108374]
<NDArray 17 @gpu(3)>
test_byteps_push_pull passed
test_byteps_trainer_param_order passed
self 201 int32 1 
[288]
<NDArray 1 @gpu(3)> 0
tensor 3 
[ 83  72   0 -36  39 -77 -22 -96 -43 -28   4  21  10 -20  75  24 -43]
<NDArray 17 @gpu(3)>
multiplied 3 
[ 332  288    0 -144  156 -308  -88 -384 -172 -112   16   84   40  -80
  300   96 -172]
<NDArray 17 @gpu(3)>
tensor before push_pull: 
[ 13.224661     0.13791323 -88.42125     25.27721     13.318825
  75.00076     74.871086   -56.566643   -52.70393     -7.9426765
 -81.11471    -57.324112    42.22908     79.71407     18.322575
   6.675613    56.61081   ]
<NDArray 17 @gpu(1)>
tensor after push_pull: 
[ -2.2648907 -26.942665   -3.1381292  30.620129   14.483914  -24.707336
  57.58606    17.05282     9.471867  -14.655745  -26.668665    4.1452713
  29.551579   31.442118   47.59712     5.5366697   6.2108374]
<NDArray 17 @gpu(1)>
test_byteps_push_pull passed
test_byteps_trainer_param_order passed
self 201 int32 1 
[288]
<NDArray 1 @gpu(1)> 0
tensor 1 
[ 83  72   0 -36  39 -77 -22 -96 -43 -28   4  21  10 -20  75  24 -43]
<NDArray 17 @gpu(1)>
multiplied 1 
[ 332  288    0 -144  156 -308  -88 -384 -172 -112   16   84   40  -80
  300   96 -172]
<NDArray 17 @gpu(1)>
tensor before push_pull: 
[-88.276085    9.469676  -28.467728   32.062294  -16.11681   -62.981438
  83.053955   19.23256    15.465963   89.03272    44.934155   -7.217002
  57.025764  -39.45068    89.96457   -55.9767      7.7130795]
<NDArray 17 @gpu(0)>
tensor after push_pull: 
[ -2.2648907 -26.942665   -3.1381292  30.620129   14.483914  -24.707336
  57.58606    17.05282     9.471867  -14.655745  -26.668665    4.1452713
  29.551579   31.442118   47.59712     5.5366697   6.2108374]
<NDArray 17 @gpu(0)>
test_byteps_push_pull passed
test_byteps_trainer_param_order passed
self 201 int32 1 
[288]
<NDArray 1 @gpu(0)> 0
tensor 0 
[ 83  72   0 -36  39 -77 -22 -96 -43 -28   4  21  10 -20  75  24 -43]
<NDArray 17 @gpu(0)>
multiplied 0 
[ 332  288    0 -144  156 -308  -88 -384 -172 -112   16   84   40  -80
  300   96 -172]
<NDArray 17 @gpu(0)>
tensor before push_pull: 
[ 21.069908 -94.97407   85.203156  52.046513  49.008537 -40.987206
  72.30864   71.563705 -17.39539  -54.30231  -87.22985   56.130363
  96.28754   70.68246   56.345104  27.549076 -98.324036]
<NDArray 17 @gpu(2)>
tensor after push_pull: 
[ -2.2648907 -26.942665   -3.1381292  30.620129   14.483914  -24.707336
  57.58606    17.05282     9.471867  -14.655745  -26.668665    4.1452713
  29.551579   31.442118   47.59712     5.5366697   6.2108374]
<NDArray 17 @gpu(2)>
test_byteps_push_pull passed
test_byteps_trainer_param_order passed
self 201 int32 1 
[288]
<NDArray 1 @gpu(2)> 0
tensor 2 
[ 83  72   0 -36  39 -77 -22 -96 -43 -28   4  21  10 -20  75  24 -43]
<NDArray 17 @gpu(2)>
multiplied 2 
[ 332  288    0 -144  156 -308  -88 -384 -172 -112   16   84   40  -80
  300   96 -172]
<NDArray 17 @gpu(2)>
[2020-01-13 06:36:30.273508: D byteps/common/global.cc:282] Shutdown BytePS: start to clean the resources (rank=0)
[2020-01-13 06:36:30.285661: D byteps/common/global.cc:282] Shutdown BytePS: start to clean the resources (rank=1)
[2020-01-13 06:36:30.306601: D byteps/common/global.cc:282] Shutdown BytePS: start to clean the resources (rank=2)
[2020-01-13 06:36:30.306861: D byteps/common/global.cc:282] Shutdown BytePS: start to clean the resources (rank=3)
[2020-01-13 06:36:30.308988: D byteps/common/shared_memory.h:45] Clear shared memory: all BytePS shared memory released/unregistered.
[2020-01-13 06:36:33.253281: D byteps/common/communicator.cc:203] listen thread joined (rank=3)
[2020-01-13 06:36:33.257222: D byteps/common/communicator.cc:203] listen thread joined (rank=3)
[2020-01-13 06:36:33.257421: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=3)
[2020-01-13 06:36:33.257491: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=3)
[2020-01-13 06:36:33.257515: D byteps/common/nccl_manager.h:63] Clear NcclManager
[2020-01-13 06:36:33.257531: D byteps/common/global.cc:335] Shutdown BytePS: all BytePS resources has been cleaned (rank=3)
[2020-01-13 06:36:33.257544: D byteps/common/operations.cc:80] BytePS has been completely shutdown now
Traceback (most recent call last):
  File "/usr/local/byteps/tests/test_mxnet.py", line 189, in <module>
    mxtest.test_byteps_push_pull_inplace()
  File "/usr/local/byteps/tests/test_mxnet.py", line 134, in test_byteps_push_pull_inplace
    incorrect results for self'
AssertionError: bps.byteps_push_pull produces                                                  incorrect results for self
[2020-01-13 06:36:33.262659: D byteps/common/shared_memory.h:45] Clear shared memory: all BytePS shared memory released/unregistered.
[2020-01-13 06:36:33.262777: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=2)
[2020-01-13 06:36:33.262825: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=2)
[2020-01-13 06:36:33.262804: D byteps/common/shared_memory.h:45] Clear shared memory: all BytePS shared memory released/unregistered.
[2020-01-13 06:36:33.262846: D byteps/common/nccl_manager.h:63] Clear NcclManager
[2020-01-13 06:36:33.262866: D byteps/common/global.cc:335] Shutdown BytePS: all BytePS resources has been cleaned (rank=2)
[2020-01-13 06:36:33.262884: D byteps/common/operations.cc:80] BytePS has been completely shutdown now
[2020-01-13 06:36:33.262841: D byteps/common/shared_memory.h:45] Clear shared memory: all BytePS shared memory released/unregistered.
[2020-01-13 06:36:33.262907: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=1)
[2020-01-13 06:36:33.262962: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=0)
[2020-01-13 06:36:33.262972: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=1)
[2020-01-13 06:36:33.262995: D byteps/common/nccl_manager.h:63] Clear NcclManager
[2020-01-13 06:36:33.263015: D byteps/common/global.cc:335] Shutdown BytePS: all BytePS resources has been cleaned (rank=1)
[2020-01-13 06:36:33.263032: D byteps/common/operations.cc:80] BytePS has been completely shutdown now
[2020-01-13 06:36:33.263031: D byteps/common/communicator.h:108] Clear BytePSCommSocket (rank=0)
[2020-01-13 06:36:33.263061: D byteps/common/nccl_manager.h:63] Clear NcclManager
[2020-01-13 06:36:33.263081: D byteps/common/global.cc:335] Shutdown BytePS: all BytePS resources has been cleaned (rank=0)
[2020-01-13 06:36:33.263099: D byteps/common/operations.cc:80] BytePS has been completely shutdown now
Traceback (most recent call last):
  File "/usr/local/byteps/tests/test_mxnet.py", line 189, in <module>
    mxtest.test_byteps_push_pull_inplace()
  File "/usr/local/byteps/tests/test_mxnet.py", line 134, in test_byteps_push_pull_inplace
    incorrect results for self'
AssertionError: bps.byteps_push_pull produces                                                  incorrect results for self
Traceback (most recent call last):
  File "/usr/local/byteps/tests/test_mxnet.py", line 189, in <module>
    mxtest.test_byteps_push_pull_inplace()
  File "/usr/local/byteps/tests/test_mxnet.py", line 134, in test_byteps_push_pull_inplace
    incorrect results for self'
AssertionError: bps.byteps_push_pull produces                                                  incorrect results for self
Traceback (most recent call last):
  File "/usr/local/byteps/tests/test_mxnet.py", line 189, in <module>
    mxtest.test_byteps_push_pull_inplace()
  File "/usr/local/byteps/tests/test_mxnet.py", line 134, in test_byteps_push_pull_inplace
    incorrect results for self'
AssertionError: bps.byteps_push_pull produces                                                  incorrect results for self
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/byteps/launcher/launch.py", line 47, in worker
    subprocess.check_call(command, env=my_env, stdout=sys.stdout, stderr=sys.stderr, shell=True)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python3 /usr/local/byteps/tests/test_mxnet.py' returned non-zero exit status 1.

Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/byteps/launcher/launch.py", line 47, in worker
    subprocess.check_call(command, env=my_env, stdout=sys.stdout, stderr=sys.stderr, shell=True)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python3 /usr/local/byteps/tests/test_mxnet.py' returned non-zero exit status 1.

Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/byteps/launcher/launch.py", line 47, in worker
    subprocess.check_call(command, env=my_env, stdout=sys.stdout, stderr=sys.stderr, shell=True)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python3 /usr/local/byteps/tests/test_mxnet.py' returned non-zero exit status 1.

Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/byteps/launcher/launch.py", line 47, in worker
    subprocess.check_call(command, env=my_env, stdout=sys.stdout, stderr=sys.stderr, shell=True)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python3 /usr/local/byteps/tests/test_mxnet.py' returned non-zero exit status 1.
```


**To Reproduce**
Steps to reproduce the behavior:
1. 
2. 
3. 
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment (please complete the following information):**

use docker/Dockerfile.mxnet

**Additional context**
Add any other context about the problem here.


___

IMO, this [line](https://github.com/bytedance/byteps/blob/master/tests/test_mxnet.py#L112) should be removed. This is because tensors are averaged by default. ([mxnet/ops.py](https://github.com/bytedance/byteps/blob/master/byteps/mxnet/ops.py#L45))

--- a/tests/test_mxnet.py
+++ b/tests/test_mxnet.py
@@ -109,7 +109,7 @@ class MXTest:
             tensor = mx.nd.random.uniform(-100, 100, shape=shapes[dim],
                                           ctx=ctx)
             tensor = tensor.astype(dtype)
-            multiplied = tensor * size
+            multiplied = tensor.copy()
             bps.byteps_declare_tensor("tensor_" + str(count))
             bps.byteps_push_pull(tensor, name= "tensor_" + str(count))
             max_difference = mx.nd.max(mx.nd.subtract(tensor, multiplied))
@@ -186,6 +186,4 @@ if __name__ == '__main__':
     mxtest.test_byteps_push_pull()
     mxtest.test_byteps_trainer_param_order()
     #mxtest.test_byteps_broadcast()
-    #mxtest.test_byteps_push_pull_inplace()
-    # TODO: fix shutdown issue https://github.com/bytedance/byteps/issues/177
-    # bps.shutdown()
+    mxtest.test_byteps_push_pull_inplace()

